overlord/udevmon.go

```diff
@@ -85,16 +87,59 @@ func (m *UDevMonitor) Connect() error {
 // handles hotplug events (devices added or removed). It returns immediately.
 // The goroutine must be stopped by calling Stop() method.
 func (m *UDevMonitor) Run() error {
+	existingDevices := make(chan *hotplug.HotplugDeviceInfo)
+	udevadmErrors := make(chan error)
+
+	if err := hotplug.EnumerateExistingDevices(existingDevices, udevadmErrors); err != nil {
+		return fmt.Errorf("failed to enumerate existing devices: %s", err)
+	}
+
 	m.tmb.Go(func() error {
+		var enumerationFinished bool
+		// Gather devices from udevadm info output (enumeration on startup) and from udev event monitor:
+		// - devices discovered on startup are reported via existingDevices channel
+		// - added/removed devices are reported via netlinkEvents channel
+		// Devices reported by netlinkEvents channel are queued until all devices from existingDevices
+		// are processed.
+		// It might happen that a device is plugged at startup and is reported by both existingDevices
+		// channel and netlinkEvents channel; such events are ignored by de-duplication logic based on device path;
+		// this de-dup logic is only applied until enumeration is finished.
 		for {
 			select {
+			case err, ok := <-udevadmErrors:
+				if ok {
+					logger.Noticef("udevadm error: %q\n", err)
+				} else {
+					udevadmErrors = nil
```

Comment:
Note it's `err, ok := <-udevadmErrors:` case for the channel. Ok is false if the channel gets closed by the sender to indicate end of data for the channel. Setting the channel to `nil` is an idiomatic way for preventing `select` statement from reporting close event for this channel anymore if the select loop is still watching other channels (it would keep reporting close of udevadmErrors forever). See e.g. https://www.godesignpatterns.com/2014/05/nil-channels-always-block.html